src/cz/crcs/ectester/standalone/libs/jni/libressl.c

```diff
@@ -0,0 +1,537 @@
+#include <openssl/opensslv.h>
+#include <openssl/objects.h>
+#include <openssl/obj_mac.h>
+#include <openssl/bn.h>
+#include <openssl/evp.h>
+#include <openssl/err.h>
+#include <openssl/ec.h>
+#include <openssl/ecdh.h>
+#include <openssl/ecdsa.h>
+
+#include "c_utils.h"
+#include "c_timing.h"
+
+
+static jclass provider_class;
+
+JNIEXPORT jobject JNICALL Java_cz_crcs_ectester_standalone_libs_LibresslLib_createProvider(JNIEnv *env, jobject self) {
+    /* Create the custom provider. */
+    jclass local_provider_class = (*env)->FindClass(env, "cz/crcs/ectester/standalone/libs/jni/NativeProvider$Libressl");
+    provider_class = (*env)->NewGlobalRef(env, local_provider_class);
+
+    jmethodID init = (*env)->GetMethodID(env, local_provider_class, "<init>", "(Ljava/lang/String;DLjava/lang/String;)V");
+
+    jstring name =  (*env)->NewStringUTF(env, LIBRESSL_VERSION_TEXT);
+    long ver_hi  = (LIBRESSL_VERSION_NUMBER & 0xff000000L) >> 28;
+    long ver_mid = (LIBRESSL_VERSION_NUMBER & 0x00ff0000L) >> 20;
+    long ver_low = (LIBRESSL_VERSION_NUMBER & 0x0000ff00L) >> 12;
+    double version = (double)ver_hi + ((double)ver_mid/10) + ((double)ver_low/100);
+
+    return (*env)->NewObject(env, provider_class, init, name, version, name);
+}
+
+JNIEXPORT void JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeProvider_00024Libressl_setup(JNIEnv *env, jobject self) {
+    OPENSSL_init_crypto(0, NULL);
+
+    INIT_PROVIDER(env, provider_class);
+
+    ADD_KPG(env, self, "EC", "Libressl");
+    ADD_KA(env, self, "ECDH", "LibresslECDH");
+    ADD_SIG(env, self, "NONEwithECDSA", "LibresslECDSAwithNONE");
+
+    init_classes(env, "Libressl");
+}
+
+JNIEXPORT jobject JNICALL Java_cz_crcs_ectester_standalone_libs_LibresslLib_getCurves(JNIEnv *env, jobject self) {
+    jclass hash_set_class = (*env)->FindClass(env, "java/util/TreeSet");
+
+    jmethodID hash_set_ctr = (*env)->GetMethodID(env, hash_set_class, "<init>", "()V");
+    jmethodID hash_set_add = (*env)->GetMethodID(env, hash_set_class, "add", "(Ljava/lang/Object;)Z");
+
+    jobject result = (*env)->NewObject(env, hash_set_class, hash_set_ctr);
+
+    size_t ncurves = EC_get_builtin_curves(NULL, 0);
+    EC_builtin_curve curves[ncurves];
+    EC_get_builtin_curves(curves, ncurves);
+
+    for (size_t i = 0; i < ncurves; ++i) {
+        jstring curve_name = (*env)->NewStringUTF(env, OBJ_nid2sn(curves[i].nid));
+        (*env)->CallBooleanMethod(env, result, hash_set_add, curve_name);
+    }
+
+    return result;
+}
+
+JNIEXPORT jboolean JNICALL Java_cz_crcs_ectester_standalone_libs_jni_NativeKeyPairGeneratorSpi_00024Libressl_keysizeSupported(JNIEnv *env, jobject self, jint keysize) {
+    size_t ncurves = EC_get_builtin_curves(NULL, 0);
+    EC_builtin_curve curves[ncurves];
+    EC_get_builtin_curves(curves, ncurves);
+
+    for (size_t i = 0; i < ncurves; ++i) {
+        EC_GROUP *curve = EC_GROUP_new_by_curve_name(curves[i].nid);
+        if (EC_GROUP_get_degree(curve) == keysize) {
```

Comment:
Is `get_degree` the correct thing in LibreSSL for getting the curve size? Even in the prime field case (`secp521r1`, etc.)?

Reply:
Per [the documentation](https://man.openbsd.org/EC_GROUP_get_degree.3),

> `EC_GROUP_get_degree()` gets the degree of the field. For _Fp_ fields this will be the number of bits in _p_. For _F2^m_ fields this will be the value _m_.

So this should be correct.